runtime/dart_isolate.cc

```diff
@@ -280,8 +283,159 @@ std::weak_ptr<DartIsolate> DartIsolate::CreateRootIsolate(
+          std::shared_ptr<DartIsolateGroupData>* unused_isolate_group_data,
+          std::shared_ptr<DartIsolate>* isolate_data, Dart_IsolateFlags* flags,
+          char** error) {
+        return Dart_CreateIsolateInGroup(
+            /*group_member=*/parent_isolate,
+            /*name=*/"PlatformIsolate",
+            /*shutdown_callback=*/
+            reinterpret_cast<Dart_IsolateShutdownCallback>(
+                DartIsolate::SpawnIsolateShutdownCallback),
+            /*cleanup_callback=*/
+            reinterpret_cast<Dart_IsolateCleanupCallback>(
+                DartIsolateCleanupCallback),
+            /*child_isolate_data=*/isolate_data,
+            /*error=*/error);
+      };
+  Dart_Isolate platform_isolate = CreateDartIsolateGroup(
+      nullptr, std::move(isolate_data), nullptr, error, isolate_maker);
+
+  Dart_EnterIsolate(parent_isolate);
+
+  if (*error) {
+    return nullptr;
+  }
+
+  if (!platform_isolate_manager->RegisterPlatformIsolate(platform_isolate)) {
+    // The PlatformIsolateManager was shutdown while we were creating the
+    // isolate. This means that we're shutting down the engine. We need to
+    // shutdown the platform isolate.
+    FML_LOG(INFO) << "Shutdown during platform isolate creation";
+    Dart_ExitIsolate();  // Exit parent_isolate.
+    Dart_EnterIsolate(platform_isolate);
+    Dart_ShutdownIsolate();
+    Dart_EnterIsolate(parent_isolate);
+    return nullptr;
+  }
+
+  Dart_EnterScope();
+  Dart_PersistentHandle entry_point_handle =
+      Dart_NewPersistentHandle(entry_point);
+  Dart_ExitScope();
+
+  platform_task_runner->PostTask([entry_point_handle, platform_isolate, port_id,
+                                  platform_isolate_manager]() {
+    if (platform_isolate_manager->IsShutdown()) {
+      // Shutdown happened in between this task being posted, and it running.
+      // platform_isolate has already been shut down. Do nothing.
+      FML_LOG(INFO) << "Shutdown before platform isolate entry point";
+      return;
+    }
+    Dart_EnterIsolate(platform_isolate);
+    Dart_EnterScope();
+    Dart_Handle entry_point = Dart_HandleFromPersistent(entry_point_handle);
+    Dart_DeletePersistentHandle(entry_point_handle);
+    Dart_Handle isolate_ready_port = Dart_NewSendPort(port_id);
+
+    // Disable Isolate.exit().
+    Dart_Handle isolate_lib = Dart_LookupLibrary(tonic::ToDart("dart:isolate"));
+    FML_CHECK(!tonic::CheckAndHandleError(isolate_lib));
+    Dart_Handle isolate_type = Dart_GetNonNullableType(
+        isolate_lib, tonic::ToDart("Isolate"), 0, nullptr);
+    FML_CHECK(!tonic::CheckAndHandleError(isolate_type));
+    Dart_Handle result =
+        Dart_SetField(isolate_type, tonic::ToDart("_mayExit"), Dart_False());
+    FML_CHECK(!tonic::CheckAndHandleError(result));
+
+    tonic::DartInvoke(entry_point, {isolate_ready_port});
+
+    if (Dart_CurrentIsolate() == nullptr) {
+      // The platform isolate entry point caused the isolate to shut down.
```

Comment:
Could you explain how this can possibly happen? If it can't, it'd be better to assert it.